core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,178 @@
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+    deleteMetadataFiles(
+        manifestPathsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest", path, exc));
+    deleteMetadataFiles(
+        manifestListsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest list", path, exc));
+  }
+
+  private List<ManifestFile> computeManifestsToDelete(
+      Set<String> manifestListsToDelete, Map<String, ManifestFile> reachableFilesAfterExpiration) {
+    Map<String, ManifestFile> manifestCandidatesToDelete =
+        computeManifestsForManifestLists(manifestListsToDelete);
+    List<ManifestFile> manifestFilesToDelete = Lists.newArrayList();
+    for (Map.Entry<String, ManifestFile> candidateManifest :
+        manifestCandidatesToDelete.entrySet()) {
+      if (!reachableFilesAfterExpiration.containsKey(candidateManifest.getKey())) {
+        manifestFilesToDelete.add(candidateManifest.getValue());
+      }
+    }
+    return manifestFilesToDelete;
+  }
+
+  private Map<String, ManifestFile> computeManifestsForManifestLists(Set<String> manifestLists) {
+    Map<String, ManifestFile> pathToManifestFiles = Maps.newHashMap();
+    for (String manifestList : manifestLists) {
+      CloseableIterable<ManifestFile> manifestFiles = readManifestFiles(manifestList);
+      for (ManifestFile manifest : manifestFiles) {
+        pathToManifestFiles.put(manifest.path(), manifest.copy());
+      }
+    }
+    return pathToManifestFiles;
+  }
+
+  // Helper to compute files to delete
+  private Set<String> findFilesToDelete(
+      List<ManifestFile> manifestFilesToDelete, List<ManifestFile> currentManifestFiles) {
+    Set<String> filesToDelete = ConcurrentHashMap.newKeySet();
+    Tasks.foreach(manifestFilesToDelete)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (ManifestReader<?> reader =
+                  ManifestFiles.open(manifest, ops.io(), ops.current().specsById())) {
+                for (ManifestEntry<?> entry : reader.entries()) {
+                  filesToDelete.add(entry.file().path().toString());
+                }
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+    Tasks.foreach(currentManifestFiles)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
```

Comment:
I think each run should check to see if there are still files to delete and skip reading the manifest if there are none.

Reply:
Fixed!